### ISSUE TITLE
When item that is closed is migrated, ADO shows: Field 'Closed Date' cannot be empty

### ISSUE THREAD

**Issue by csaket** [solidify/jira-azuredevops-migrator#793]
**Describe the problem**
Migration of a Bug with status Closed from Jira Server 8.6 to ADO, workitem of type Bug created correctly, with correct State of Closed.
Issue is that ADO now gives this error: 
Field 'Closed Date' cannot be empty.

Is there anything special that I might have missed out? I am using mapping to a CMMI based process.
For Closed, the status to State mapping is just:

            {
              "source": "Closed",
              "target": "Closed"
            },

**Comment by Alexander-Hjelm** [comment]
This should work fine, and has worked fine for each of my migrations, so most likely it is due to some configuration in your ADO process model.

Have you modified the Bug process or made any changes to the XML model?

**Comment by jus062** [comment]
I noticed this behaviour today as well, but am getting it for all Closed items when using v3.0.129

I've run a test using our old version 2.3.117 and it works correctly.

**Comment by Alexander-Hjelm** [comment]
Strange, because it works just fine in our tests.

@jus062 are you also using a CMMI-based model like @csaket ?

**Comment by jus062** [comment]
No, I'm using an inherited agile process. We do have rules on Tasks and Stories, but I disabled those before migrating.

**Comment by jus062** [comment]
Ok, I just reran one test work item on v3.0.129 and this is what I see 

<img width="305" alt="image" src="https://github.com/solidify/jira-azuredevops-migrator/assets/94574483/8a68cfa1-4e8d-400a-8c3f-6e3f2a1a4b1b">

and this is from the json

      "Time": "2023-06-05T21:42:50.758+10:00",
      "Index": 12,
      "Fields": [
        {
          "ReferenceName": "System.State",
          "Value": "Closed"
        }
      ],

**Comment by Alexander-Hjelm** [comment]
Ok! marking ab bug

**Comment by RoyalWulf** [comment]
Hi Getting this issue as well.
I don't have a closed date on any closed work items
I don't have an activated date on any work items either. Should I on Active work items?

**Comment by Alexander-Hjelm** [comment]
@UweStein , @RoyalWulf , this issue comes from a bug in the tool. The field `Microsoft.VSTS.Common.ClosedDate` needs to be set to the revision's `Time`. Otherwise the ADO Work Item form GUI will complain.

This is an issue when using the WIT Rest API with the `Bypass rules on work item updates` flag set to `true`.

I will work on a fix and release anew, most likely during this week or the next!

**Comment by Alexander-Hjelm** [comment]
See WiClientUtils.CorrectClosedByAndClosedDate, lines 625 - 644

**Comment by RoyalWulf** [comment]
Hi,
When are you going to make a new release?
There are several bugs (including this one) that are fixed and we would really like the fixes.
Cheers,
John


_____________________________________________
John Edwards

Agile Enablement Lead

Digital Services

Te Ratonga Tautoko Hangarau

T: +64 3 369 3459 (Internal 93459)

University of Canterbury | Te Whare Wananga o Waitaha

Private Bag 4800, Christchurch 8140, New Zealand

_____________________________________________



________________________________

**Comment by Alexander-Hjelm** [comment]
@RoyalWulf, I have released a new version (3.0.159), but seems like we are dealing with permission issues. I am unable to promote v3.0.159 to the latest release. In the meantime you can download it here: https://github.com/solidify/jira-azuredevops-migrator/releases/tag/v3.0.159

**Comment by Alexander-Hjelm** [comment]
@RoyalWulf I have now promoted v3.0.159 to be the latest release :)

**Comment by eplnde** [comment]
@Alexander-Hjelm After this fix I seem to get a "The value for field 'Closed Date' must be empty." after migrating bugs that have be reopened (had their state changed from Closed to New)

**Comment by Alexander-Hjelm** [comment]
@eplnde , Ok! I'm reopening this issue. I'll try and look at it ASAP.

**Comment by surendra-github** [comment]
Thank you so much !!

**Comment by Alexander-Hjelm** [comment]
To reproduce the error:
1. Create a bug in Jira with the To Do status
2. Move the bug to Closed
3. Reopen the bug (move to active)
4. Run an export
5. Run an import

Here is a sample workitem .json file that successfully recreates the error:

```json
{
  "Type": "Bug",
  "OriginId": "AGILEDEMO-34",
  "WiId": -1,
  "Revisions": [
    {
      "Author": "alexander.hjelm@solidify.dev",
      "Time": "2023-10-07T09:03:50.441+02:00",
      "Index": 0,
      "Fields": [
        {
          "ReferenceName": "System.Title",
          "Value": "[AGILEDEMO-34] (Jira smoke tests) bug with status closed"
        },
        {
          "ReferenceName": "Microsoft.VSTS.Common.Priority",
          "Value": "3"
        },
        {
          "ReferenceName": "Custom.Reference",
          "Value": "AGILEDEMO-34"
        },
        {
          "ReferenceName": "System.State",
          "Value": "New"
        }
      ],
      "Links": [],
      "Attachments": [],
      "Commit": null,
      "AttachmentReferences": false
    },
    {
      "Author": "alexander.hjelm@solidify.dev",
      "Time": "2023-10-07T09:05:04.099+02:00",
      "Index": 1,
      "Fields": [
        {
          "ReferenceName": "System.AssignedTo",
          "Value": "alexander.hjelm@solidify.dev"
        },
        {
          "ReferenceName": "System.State",
          "Value": "Active"
        }
      ],
      "Links": [],
      "Attachments": [],
      "Commit": null,
      "AttachmentReferences": false
    },
    {
      "Author": "alexander.hjelm@solidify.dev",
      "Time": "2023-10-07T09:05:10.141+02:00",
      "Index": 2,
      "Fields": [
        {
          "ReferenceName": "System.State",
          "Value": "Closed"
        }
      ],
      "Links": [],
      "Attachments": [],
      "Commit": null,
      "AttachmentReferences": false
    },
    {
      "Author": "alexander.hjelm@solidify.dev",
      "Time": "2023-10-07T09:07:09.779+02:00",
      "Index": 3,
      "Fields": [
        {
          "ReferenceName": "System.State",
          "Value": "Active"
        }
      ],
      "Links": [],
      "Attachments": [],
      "Commit": null,
      "AttachmentReferences": false
    }
  ]
}
```

**Comment by Alexander-Hjelm** [comment]
PR created with fix. Please try the feature branch and let me know the results!

**Comment by Alexander-Hjelm** [comment]
Merged #878. Reopen this issue if the problem existsi.
